tests/test_simpleGenericMap.py

```diff
@@ -216,6 +218,148 @@ def testClear(self):
+
+    def _checkCppUpdates(self, testmap, msg=''):
+        for key, value in self.data.items():
+            self.assertIn(key, testmap, msg=msg)
+            self.assertEqual(value, testmap[key], msg='key=' + key + ', ' + msg)
+            cppLib.assertKeyValue(testmap, key, value)
+        cppLib.makeCppUpdates(testmap)
+
+        for key, value in {'answer': 42, 'pi': 3.0, 'string': False}.items():
+            # Test both Python and C++ state
+            self.assertIn(key, testmap, msg=msg)
+            self.assertEqual(value, testmap[key], msg='key=' + key + ', ' + msg)
+            cppLib.assertKeyValue(testmap, key, value)
+
+    def testCppUpdates(self):
+        """Check that changes to built-in types made in C++ are visible in
+        both languages.
+        """
+        self._checkCppUpdates(self.pymap, msg='map=pymap')
+        self._checkCppUpdates(self.cppmap, msg='map=cppmap')
+
+    def _checkPythonStorableUpdates(self, testmap, msg=''):
+        cppLib.addCppStorable(testmap)
+        self.assertIn('cppValue', testmap, msg=msg)
+        self.assertEqual(testmap['cppValue'], cppLib.CppStorable('value'), msg=msg)
+        self.assertIn('cppPointer', testmap, msg=msg)
+        self.assertEqual(testmap['cppPointer'], cppLib.CppStorable('pointer'), msg=msg)
+
+        # should have no effect because pybind11 copies Storable values for safety
+        testmap['cppValue'].value = 'new_value'
+        testmap['cppPointer'].value = 'extra_pointy'
+
+        for key, value in {'cppValue': cppLib.CppStorable('value'),
+                           'cppPointer': cppLib.CppStorable('extra_pointy'),
+                           }.items():
+            # Test both Python and C++ state
+            self.assertIn(key, testmap, msg=msg)
+            self.assertEqual(value, testmap[key], msg='key=' + key + ', ' + msg)
+            cppLib.assertKeyValue(testmap, key, value)
+
+    def testPythonStorableUpdates(self):
+        """Check that changes to Storables made in Python are visible in
+        both languages.
+        """
+        self._checkPythonStorableUpdates(self.pymap, msg='map=pymap')
+        self._checkPythonStorableUpdates(self.cppmap, msg='map=cppmap')
+
+    def _checkCppStorableRead(self, testmap, msg=''):
+        # WARNING: the Python variables holding PyStorable must survive to the end of the test
+        # This is a known bug in pybind11; see DM-21314
+        storableData = {'answer': PyStorable(42),
+                        'question': PyStorable('Unknown'),
+                        }
+        testmap.update(storableData)
+
+        for key, value in storableData.items():
+            self.assertIn(key, testmap, msg=msg)
+            self.assertEqual(value, testmap[key], msg='key=' + key + ', ' + msg)
+            # Exercise C++ equality operator
+            cppLib.assertKeyValue(testmap, key, PyStorable(value.value))
+            # TODO: move to test_Storable.py
```

Comment:
Was this TODO note intended for this ticket or a future one?

Reply:
This one. I could swear I removed it in 51b1cdaae25e7e80fa500bbf465c01a72345416d... :confused: